code/modules/shuttle/supply.dm

```diff
@@ -231,365 +239,338 @@ GLOBAL_LIST_EMPTY(exports_types)
+/datum/supply_ui/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	if(..())
+		return
+	switch(action)
+		if("cart")
+			var/datum/supply_packs/P = SSpoints.supply_packs[text2path(params["id"])]
+			if(!P)
+				return
+			var/shopping_cart = get_shopping_cart(ui.user)
+			switch(params["mode"])
+				if("removeall")
+					shopping_cart -= P.type
+				if("removeone")
+					if(shopping_cart[P.type] > 1)
+						shopping_cart[P.type]--
 					else
-						dat += "*ASRS is busy*"
-					dat += "<BR>\n<BR>"
-				else
-					dat += "Lowered<BR>"
-					if (SSshuttle.supply.mode == SHUTTLE_IDLE)
-						dat += "<A href='?src=\ref[src];send=1'>Raise platform</A>"
-					else if (SSshuttle.supply.mode == SHUTTLE_IGNITING)
-						dat += "<A href='?src=\ref[src];cancel_send=1'>Cancel</A>"
+						shopping_cart -= P.type
+				if("addone")
+					if(shopping_cart[P.type])
+						shopping_cart[P.type]++
 					else
-						dat += "*ASRS is busy*"
-					dat += "<BR>\n<BR>"
-
-
-		dat += {"<HR>\nSupply points: [round(SSpoints.supply_points)]<BR>\n<BR>
-		\n<A href='?src=\ref[src];order=categories'>Order items</A><BR>\n<BR>
-		\n<A href='?src=\ref[src];viewrequests=1'>View requests</A><BR>\n<BR>
-		\n<A href='?src=\ref[src];vieworders=1'>View orders</A><BR>\n<BR>"}
-
-
-	var/datum/browser/popup = new(user, "computer", "<div align='center'>Automated Storage and Retrieval System</div>", 575, 450)
-	popup.set_content(dat)
-	popup.open(FALSE)
-	onclose(user, "computer")
+						shopping_cart[P.type] = 1
+				if("addall")
+					var/cart_cost = 0
+					for(var/i in shopping_cart)
+						var/datum/supply_packs/SP = SSpoints.supply_packs[i]
+						cart_cost += SP.cost * shopping_cart[SP.type]
+					var/excess_points = SSpoints.supply_points - cart_cost
+					var/number_to_buy = round(excess_points / P.cost)
+					if(shopping_cart[P.type])
+						shopping_cart[P.type] += number_to_buy
+					else
+						shopping_cart[P.type] = number_to_buy
+			. = TRUE
+		if("send")
+			if(SSshuttle.supply.mode == SHUTTLE_IDLE && is_mainship_level(SSshuttle.supply.z))
+				if (!SSshuttle.supply.check_blacklist())
+					to_chat(usr, "For safety reasons, the Automated Storage and Retrieval System cannot store live, non-xeno organisms, classified nuclear weaponry or homing beacons.")
+					playsound(SSshuttle.supply.return_center_turf(), 'sound/machines/buzz-two.ogg', 50, 0)
+				else
+					playsound(SSshuttle.supply.return_center_turf(), 'sound/machines/elevator_move.ogg', 50, 0)
+					SSshuttle.moveShuttle("supply", "supply_away", TRUE)
+			else
+				var/obj/docking_port/D = SSshuttle.getDock("supply_home")
+				playsound(D.return_center_turf(), 'sound/machines/elevator_move.ogg', 50, 0)
+				SSshuttle.moveShuttle("supply", "supply_home", TRUE)
+			. = TRUE
+		if("approve")
+			var/datum/supply_order/O = SSpoints.requestlist["[params["id"]]"]
+			if(!O)
+				O = SSpoints.deniedrequests["[params["id"]]"]
+			if(!O)
+				return
+			SSpoints.approve_request(O, ui.user)
+			. = TRUE
```

Comment:
Does TGUI handle validation of the request?
How do we stop some random href validating these?